### ISSUE TITLE
Help in getting a custom IServiceProvider working with the NodaTime NuGet package.

### ISSUE THREAD

**Issue by StevenRasmussen** [dotnet/efcore#27566]
Preface: I am the author of the [EF Core NodaTime](https://www.nuget.org/packages/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime) NuGet package whose project is located here: [EFCore.SqlServer.NodaTime](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime).

I originally had an issue which I ran into when upgrading my project to support .Net 6 (see [here](https://github.com/dotnet/efcore/issues/26122)).  Through some assistance I was able to get it working, which it does as long as you use the default `IServiceProvider`.

A user is running into an issue when registering a custom `IServiceProvider` (ticket [here](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/issues/27)). I have been able to reproduce the issue and have created a new branch and added a unit test to help with debugging the issue.
* Branch is [here](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/tree/custom-service-collection-issue)
* Unit test is [here](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/blob/custom-service-collection-issue/src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.Tests/ServiceCollectionTest.cs#[L11-L26])

The exception that is thrown is:
```
Message: 
    System.ArgumentNullException : Value cannot be null. (Parameter 'provider')

  Stack Trace: 
    ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
    DbContextServices.CreateModel(Boolean designTime)
    DbContextServices.get_Model()
    <>c.<TryAddCoreServices>b__8_4(IServiceProvider p)
    CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
    CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
    CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
    CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
    <>c__DisplayClass2_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
    ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
    ServiceProviderEngineScope.GetService(Type serviceType)
    ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
    NodaTimeOptionsExtension.Validate(IDbContextOptions options) line 34
    ServiceProviderCache.ValidateOptions(IDbContextOptions options)
    ServiceProviderCache.GetOrAdd(IDbContextOptions options, Boolean providerRequired)
    DbContext.ctor(DbContextOptions options)
    RacingContext.ctor(DbContextOptions options) line 15
    ServiceCollectionTest.CustomServiceCollection_ShouldNotThrowError() line 25
```

Similar to my original ticket above, I'm not sure if there is a bug in EFCore or if there is something wrong with how the services are being registered [here](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/blob/custom-service-collection-issue/src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Extensions/ServiceCollectionExtensions.cs#[L13-L20])

If I comment out the registration of any of the `IMethodCallTranslatorPlugin` (lines [15-20](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/blob/custom-service-collection-issue/src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Extensions/ServiceCollectionExtensions.cs#[L15-L20])) then the exception does not occur.

I apologize for reaching out here for support on this as I know your time is valuable but I'm at a loss on this one.  I've tried debugging with the source symbols downloaded and navigating through the EF code but without any luck as it's a little over my head how the internal registration works.  I appreciate any assistance with this.

**Comment by ajcvickers** [comment]
@StevenRasmussen Looks like the issue discussed in #27276.

Breaking change docs for this are here: https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-6.0/breaking-changes#query-services

Let us know if this doesn't help. Appreciate your work on the Noda time plugin. 👏 👏 👏

**Comment by StevenRasmussen** [comment]
Thanks for the quick response!  Unfortunately, it doesn't help :(  The documentation suggests using `TryAdd`, which is what I had to change it to in order to get it to work with .Net 6.  Unfortunately, the error still throws when using a custom `IServiceProvider` and calling `UseInternalServiceProvider`.  I also tried the suggestion of using `AddScoped` to register the services, also to no avail.
